File: vi/codes/ruby/chapter_backtracking/preorder_traversal_ii_compact.rb
=begin
File: preorder_traversal_ii_compact.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

### Duyệt trước: ví dụ 2 ###
def pre_order(root)
  return unless root

  # Thử
  $path << root

  # Ghi lại lời giải
  $res << $path.dup if root.val == 7

  pre_order(root.left)
  pre_order(root.right)

  # Quay lui
  $path.pop
end

### Mã chương trình chính ###
if __FILE__ == $0
  root = arr_to_tree([1, 7, 3, 4, 5, 6, 7])
  puts "\nInitialize binary tree"
  print_tree(root)

  # Duyệt trước
  $path, $res = [], []
  pre_order(root)

  puts "\nOutput all paths from root node to node 7"
  for path in $res
    p path.map { |node| node.val }
  end
end

File: vi/codes/typescript/chapter_backtracking/preorder_traversal_ii_compact.ts
/**
 * File: preorder_traversal_ii_compact.ts
 * Created Time: 2023-05-09
 * Author: Justin (xiefahit@gmail.com)
 */

import { type TreeNode } from '../modules/TreeNode';
import { arrToTree } from '../modules/TreeNode';
import { printTree } from '../modules/PrintUtil';

/* Duyệt trước (preorder): Ví dụ 2 */
function preOrder(
    root: TreeNode | null,
    path: TreeNode[],
    res: TreeNode[][]
): void {
    if (root === null) {
        return;
    }
    // Thử
    path.push(root);
    if (root.val === 7) {
        // Ghi lại lời giải
        res.push([...path]);
    }
    preOrder(root.left, path, res);
    preOrder(root.right, path, res);
    // Quay lui
    path.pop();
}

// Driver Code
const root = arrToTree([1, 7, 3, 4, 5, 6, 7]);
console.log('\nInitialize binary tree');
printTree(root);

// Duyệt trước (preorder)
const path: TreeNode[] = [];
const res: TreeNode[][] = [];
preOrder(root, path, res);

console.log('\nOutput all paths from root node to node 7');
res.forEach((path) => {
    console.log(path.map((node) => node.val));
});

export {};
